boards: Add TF-M support for all nordic NS boards

joerchan opened the pull request:
Add generic SoC support for the supported nordic SoCs:
- nrf5340
- nrf9160
- nrf9120

Add generic SoC support by taking board specific configurations from
zephyr devicetree and kconfig.

joerchan committed:
boards: Enable TF-M by default for nordic SoC NS boards

Enable TF-M by default for nordic nRF5340 and nrf9160 SoC boards.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
tfm: nordic_nrf: Add generic SoC support for nordic SoCs

Add generic SoC support for the supported nordic SoCs:
- nrf5340
- nrf9160
- nrf9120

Add generic SoC support by taking board specific configurations from
zephyr devicetree and kconfig.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
doc: Remove note for missing tf-m/ns target support

All nRF91 and nRF5340 boards (except the DKs) should support TF-M/_ns
now.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

gmarull commented on modules/trusted-firmware-m/nordic_nrf/include/util/array.h:
```diff
@@ -0,0 +1,22 @@
+extern "C" {
+#endif
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(arr) (sizeof(arr)/sizeof(arr[0]))
+#endif
```
just include zephyr util.h

gmarull commented on boards/arm/actinius_icarus/Kconfig.defconfig:
```diff
@@ -8,6 +8,21 @@ if BOARD_ACTINIUS_ICARUS || BOARD_ACTINIUS_ICARUS_NS
+# By default, if we build with TF-M, instruct build system to
+# flash the combined TF-M (Secure) & Zephyr (Non Secure) image
+config TFM_FLASH_MERGED_BINARY
+	bool
```
bool redundant? should not be defined here

gmarull reviewed (dismissed):
for 2nd commit: there's no need to add Nordic internals (NCSDK-14811) to upstream commits.

erwango reviewed (commented):
For commit: `boards: Enable TF-M by default for nordic SoC NS boards`: 
What about adding a Kconfig TFM_NS_TARGET set to y by _NS targets which would allow to do all this at a single location ?

joerchan committed:
boards: Add uart1 pinctrl assignment to nrf5340 Audio DK

Add uart1 pinctrl assignment to nrf5340 Audio DK.
Assigned according to Schematic for this board.

Uart1 pins needed by TF-M when using the NS board variant.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@RaytacStanley The RayTac DKs mdbt53_db_40 and mdbt53v_db_40 don't have a uart1 node with pins.
Is it possible to add one and which pins can I assign? It's possible to just add the RX and TX pins.

joerchan committed:
boards: raytac_mdbt53_db_40_nrf5340: Add HW flow control pins

Add HW flow control pins for raytac_mdbt53_db_40_nrf5340 board.
Use the same pins as the cpunet has been assigned, as the RX / TX lines
was also using the same as cpunet.
Remove bias-pull-up as uart1 using same pins on cpunet does not have
this either.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
boards: raytac_mdbt53v_db_40_nrf5340: Add uart1 pin assignment

Add uart1 pin assignment for the raytac_mdbt53v_db_40_nrf5340 board.
This is required in order to support building TF-M for the NS variant.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

RaytacStanley commented:
@joerchan It is recommended that you change it to the following setting:
raytac_mdbt53_db_40_nrf5340 :
UART1 RX: P1.00
UART1 TX: P1.01
UART1 RTS: P1.14
UART1 CTX: P1.15
![image](https://github.com/zephyrproject-rtos/zephyr/assets/107376255/26a1aca9-c6ec-4ccb-aa97-54fc32138239)

raytac_mdbt53v_db_40_nrf5340 :
UART1 RX: P0.04
UART1 TX: P0.05
UART1 RTS: P1.02
UART1 CTX: P1.03
![image](https://github.com/zephyrproject-rtos/zephyr/assets/107376255/0761d79d-339e-4246-94d1-7b825a74cd42)

joerchan commented:
@RaytacStanley Thanks for the information.

I have noticed that:
For raytac_mdbt53v_db_40_nrf5340,  P1.02 and P1.03 are also used for the i2c1 node.
Maybe P0.25 and P0.26 can be used, or is this not an issue?

For raytac_mdbt53_db_40_nrf5340, P1.14 and P1.15 are also used for the spi3 node.
Maybe P0.11 and P0.12 can be used, or is this not an issue?

Note that TF-M can work without the RTS and CTS pins also.

Does any of the pins require the bias-pull-up configuration?
I am seeing a bit mix of this configuration usage.
For raytac_mdbt53_db_40_nrf5340 cpuapp and cpunet is having different bias-pull-up config for TX and RX lines on the uart node, even though it's the same pins.

anangl commented on modules/trusted-firmware-m/nordic_nrf/include/RTE_Device.h:
```diff
@@ -0,0 +1,53 @@
+/* ARRAY_SIZE causes a conflict as it is defined both by TF-M and indirectly by devicetree.h */
+#undef ARRAY_SIZE
+#include <zephyr/devicetree.h>
+#include <nrf-pinctrl.h>
```
Is this `zephyr/dt-bindings/pinctrl/nrf-pinctrl.h`? Perhaps I'm missing something, but is it actually needed to include it here?

joerchan replied:
I think that is true, it's only needed for the vanilla TF-M RTE_Device.h headers, which don't have access to this header.
Will remove this include.

anangl commented on soc/arm/nordic_nrf/Kconfig:
```diff
@@ -25,6 +25,13 @@ config NRF_SOC_SECURE_SUPPORTED
 	  by the secure firmware.
 
+config TFM_LOG_LEVEL_SILENCE
+	default y if !$(dt_nodelabel_has_prop,uart1,pinctrl_names)
```
```suggestion
	default y if !$(dt_nodelabel_has_prop,uart1,pinctrl-names)
```

erwango commented:
> > For commit: `boards: Enable TF-M by default for nordic SoC NS boards`: What about adding a Kconfig TFM_NS_TARGET set to y by _NS targets which would allow to do all this at a single location ?
> 
> @erwango I have done something that addresses the concern of a single location, but limited to the nordic SOC family. This means that there is no additional kconfig added, and that building for nordic nonsecure boards results in enabling TF-M by default.

Ok, fine with me. Out of curiosity, did you had a specific concern for limit this to NRF ?

joerchan commented on modules/trusted-firmware-m/nordic_nrf/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+#
+# Copyright (c) 2023, Nordic Semiconductor ASA.
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
Yeah, I guess so. will update

anangl commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -21,6 +18,9 @@ config TFM_BOARD
 	default "arm/musca_b1" if BOARD_MUSCA_B1
 	default "arm/musca_s1" if BOARD_MUSCA_S1
 	default "lairdconnectivity/bl5340_dvk_cpuapp" if BOARD_BL5340_DVK_CPUAPP_NS
+	default "${ZEPHYR_BASE}/modules/trusted-firmware-m/nordic_nrf/nrf9160" if SOC_NRF9160
```
ah, right, I confused those two

joerchan commented:
> Ok, fine with me. Out of curiosity, did you had a specific concern for limit this to NRF ?

I just don't want to make this choice for everyone. For nordic I know what the policy is, for the others I have no idea.

anangl commented on modules/trusted-firmware-m/nordic_nrf/include/device_cfg.h:
```diff
@@ -0,0 +1,32 @@
+
+#define DEFAULT_UART_BAUDRATE DT_PROP_OR(DT_NODELABEL(TFM_UART), current_speed, 115200)
+
+#if DT_PROP_OR(DT_NODELABEL(TFM_UART), hw_flow_control, 0) == 1
```
nit: `DT_NODE_HAS_PROP()` always returns 1 for boolean properties, so this will anyway call `DT_PROP()`
```suggestion
#if DT_PROP(DT_NODELABEL(TFM_UART), hw_flow_control)
```

joerchan replied:
Not sure I understood the argument, DT_NODE_HAS_PROP is not used here, but I can see that they behave the same so I will take your suggestion.

anangl replied:
`DT_PROP_OR()` calls `DT_NODE_HAS_PROP()` internally, and if that returns 1, `DT_PROP()` is called.
